lightning/src/chain/channelmonitor.rs

```diff
@@ -2869,7 +2869,37 @@ impl<Signer: Sign> ChannelMonitorImpl<Signer> {
 
 		let mut watch_outputs = Vec::new();
 		let mut claimable_outpoints = Vec::new();
-		for tx in &txn_matched {
+		'tx_iter: for tx in &txn_matched {
+			let txid = tx.txid();
+			// If a transaction has already been confirmed, ensure we don't bother processing it duplicatively.
+			if Some(txid) == self.funding_spend_confirmed {
+				log_debug!(logger, "Skipping redundant processing of funding-spend tx {} as it was previously confirmed", txid);
+				continue 'tx_iter;
+			}
+			for ev in self.onchain_events_awaiting_threshold_conf.iter() {
+				if ev.txid == txid {
+					if let Some(conf_hash) = ev.block_hash {
+						assert_eq!(header.block_hash(), conf_hash,
+							"Transaction {} was already confirmed and is being re-confirmed in a different block.\n\
+							This indicates a severe bug in the transaction connection logic - a reorg should have been processed first!", ev.txid);
```

Comment:
Hmm, if we don't panic here then how can we rather ensure that the client returns to a correct view of the chain by first getting them to call `transaction_unconfirmed`? I'm not sure what to do in this more serious case.

Reply:
> Yea, we could skip it and just return, but I'm not sure what that really means for downstream devs - they'll presumably just keep going and we'll just reject all their chain events, which isn't right either. Ultimately a bug in the chain-connection logic is an absolutely critical bug in a downstream application, so panic seems somewhat warranted.

Alright, then agreed that panicking is the way to go here *if* we want to treat this as a hard error. But can't we just call `transaction_unconfirmed` for them and therefore relax the strict ordering requirement that it has to be called before `transactions_confirmed`?

Reply:
Mh, yes, that's something to consider, but we won't get around requiring that users confirm transactions according to the 'topological ordering'. But if we'd unconfirm instead of panicking, we'd make the interface just a tad more forgiving.

Btw. I'm not sure how users *should* handle the case you just described post-#1846? Wouldn't they have the exact same issue when they try to unconfirm the first transaction themselves?

Reply:
Ah, true, it comes down to a proper ordering of the `Confirm` calls. Uff, it's really narrow how this interface can be used without shooting off a foot. Will make sure to cover this case in upcoming when working on docs via #1188.